test/QueryPlannerTest.cpp

```diff
@@ -2728,6 +2731,120 @@ TEST(QueryPlanner, FilterIsNotRewritten) {
+                          scan("?x", "<p>", "?y"), scan("?a", "<p>", "?b")),
+                      scan("?m", "<p>", "?n")),
+                  h::spatialJoinFilterSubstitute(
+                      -1, -1, V{"?y"}, V{"?b"}, std::nullopt,
+                      PayloadVariables::all(), algo, sjType1,
+                      h::spatialJoinFilterSubstitute(
+                          -1, -1, V{"?y"}, V{"?n"}, std::nullopt,
+                          PayloadVariables::all(), algo, sjType2,
+                          scan("?x", "<p>", "?y"), scan("?m", "<p>", "?n")),
+                      scan("?a", "<p>", "?b"))));
+  }
+
+  // Two geo relation filters on the same variables: The second one may not be
+  // substituted by a spatial join as this spatial join would be incomplete
+  // (that is: have only one child).
+  h::expect(
+      "PREFIX geof: <http://www.opengis.net/def/function/geosparql/> "
+      "SELECT * WHERE {"
+      "?a <p> ?b ."
+      "?m <p> ?n ."
+      "FILTER geof:sfCovers(?n, ?b) ."
+      "FILTER geof:sfContains(?n, ?b) .  }",
+      ::testing::AnyOf(
+          h::Filter("geof:sfCovers(?n, ?b)",
+                    h::spatialJoinFilterSubstitute(
+                        -1, -1, V{"?n"}, V{"?b"}, std::nullopt,
+                        PayloadVariables::all(), algo, CONTAINS,
+                        scan("?m", "<p>", "?n"), scan("?a", "<p>", "?b"))),
+          h::Filter("geof:sfContains(?n, ?b)",
+                    h::spatialJoinFilterSubstitute(
+                        -1, -1, V{"?n"}, V{"?b"}, std::nullopt,
+                        PayloadVariables::all(), algo, COVERS,
+                        scan("?m", "<p>", "?n"), scan("?a", "<p>", "?b")))));
+
+  // Combination of geo relation filter and geo distance filter
+  h::expect(
+      "PREFIX geof: <http://www.opengis.net/def/function/geosparql/> "
+      "SELECT * WHERE {"
+      "?a <p> ?b ."
+      "?x <p> ?y ."
+      "?m <p> ?n ."
+      "FILTER(geof:metricDistance(?b, ?y) <= 1000) ."
+      "FILTER geof:sfContains(?n, ?b) .  }",
+      ::testing::AnyOf(
+          h::spatialJoinFilterSubstitute(
+              1000, -1, V{"?b"}, V{"?y"}, std::nullopt, PayloadVariables::all(),
+              algo, WITHIN_DIST,
+              h::spatialJoinFilterSubstitute(
+                  -1, -1, V{"?n"}, V{"?b"}, std::nullopt,
+                  PayloadVariables::all(), algo, CONTAINS,
+                  scan("?m", "<p>", "?n"), scan("?a", "<p>", "?b")),
+              scan("?x", "<p>", "?y")),
+          h::spatialJoinFilterSubstitute(
+              -1, -1, V{"?n"}, V{"?b"}, std::nullopt, PayloadVariables::all(),
+              algo, CONTAINS, scan("?m", "<p>", "?n"),
+              h::spatialJoinFilterSubstitute(
+                  1000, -1, V{"?b"}, V{"?y"}, std::nullopt,
+                  PayloadVariables::all(), algo, WITHIN_DIST,
+                  scan("?a", "<p>", "?b"), scan("?x", "<p>", "?y")))));
+}
```

Comment:
Add the case for the two variables being the same (can this maybe throw immediately, as it currently doesn't really make sense).